### ISSUE TITLE
ofParameter::makeReferenceTo introduces problematic event listener behavior

### ISSUE THREAD

**Issue by bakercp** [openframeworks/openFrameworks#4676]
I would like to be able to figure out which ofParameter is "sending" a value using the classic event syntax (e.g. `void onFloatParameterWithSender(const void* sender, float& value);`), but private internal value changed pointer means that the source of the data can be confusing ...

```cpp
...
    void onFloatParameterWithSender(const void* sender, float& value);

    ofParameter<float> parameter0;
    ofParameter<float> parameter1;
```

And a setup like this:

```cpp
void ofApp::setup()
{
    // set up a listener for parameter 0 
    // in the callback (sender == &parameter0) is true
    parameter0.addListener(this, &ofApp::onFloatParameterWithSender);
   
    // set up a listener for parameter 1
    // in the callback (sender == &parameter1) is true
    parameter1.addListener(this, &ofApp::onFloatParameterWithSender);

    // make 
    parameter1.makeReferenceTo(parameter0);
    // this is a no-op because this listener already exists.
    parameter1.addListener(this, &ofApp::onFloatParameterWithSender);

    // now, in the callback (sender == &parameter1) will never be true because it is a reference
    // to parameter0's internal data.
}

```

Perhaps I'm just not thinking about this the right way, but I'd propose that we need to document the internal workings of ofParameter better or figure out a way to better control the fact that `makeReferenceTo` will make listeners quietly disappear when their internal data is replaced by the target's shared data.

Currently, if I want to bind a parameter to my gui (and make it appear that a gui widget is the source of the data), I have to add a layer of abstraction which automatically subscribes to the events of the newly referred parameter and then forwards them with the appropriate gui-sender pointer.

**Comment by frauzufall** [comment]
I also wanted to bind my gui dynamically to parameters and tried that with `makeReferenceTo` for a while because it seems to be handy but I ended up with the similar solution you did. Would be great to have an easier way to bind parameters to each other.

**Comment by arturoc** [comment]
yes this is kind of strange but in general having to rely on a void* to get the sender is pretty old school anyway, there's a PR to register lambdas... where that won't even make sense any more in some cases and the events are registered to a numeric id.

Perhaps we should change the signature for an ofParameter listener get the parameter as a sender and add a comparison method to check which parameter is sending:

```cpp
void onChanged(ofParameter<float> & p){
    p.is( parameter1 );
    p.is( parameter0 );
} 
```

in your example both should be true.

**Comment by bakercp** [comment]
Hi @arturoc, @frauzufall and I recently chatted about this in Denver and it's a little bit of a hold-up for various GUI efforts.  Do you have any further thoughts on the right approach here?  Your proposed solution seems reasonable to me and I'm wondering how it might mesh with the new `ofEventListener` syntax.  Thanks!

**Comment by arturoc** [comment]
i'll look into this at some point but don't think it's holding up anything it's just a pattern that some people might use but not very related with the planed changes for the gui if i remember well.

i think what i proposed could work but haven't had time to test it yet.

The new ofEventListener just simplifies the registering of events but is not very related with this.
